core/src/main/java/org/elasticsearch/script/SearchScript.java

```diff
@@ -19,30 +19,145 @@
 package org.elasticsearch.script;
 
 import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.Scorer;
+import org.apache.lucene.util.SetOnce;
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.common.lucene.ScorerAware;
+import org.elasticsearch.search.lookup.LeafDocLookup;
+import org.elasticsearch.search.lookup.LeafSearchLookup;
 import org.elasticsearch.search.lookup.SearchLookup;
 
 import java.io.IOException;
 import java.util.Map;
 
 /**
- * A search script.
+ * A generic script used for per document use cases.
+ *
+ * Using a {@link SearchScript} works as follows:
+ * <ol>
+ *     <li>Constructing using a {@link Factory}</li>
+ *     <li>Creating a duplicate {@link SearchScript} bound to a segment using {@link #forSegment(LeafReaderContext)}</li>
+ *     <li>Calling {@link #setDocument(int)} to indicate which document in the segment the script should be run for next</li>
+ *     <li>Calling one of the {@code run} methods: {@link #run()}, {@link #runAsDouble()}, or {@link #runAsLong()}</li>
+ * </ol>
  */
-public interface SearchScript {
+public abstract class SearchScript implements ScorerAware, ExecutableScript, Cloneable {
+
+    /** The generic runtime parameters for the script. */
+    public final Map<String, Object> params;
+
+    /** A lookup for the index this script will operate on. */
+    public final SearchLookup lookup;
+
+    /** A leaf lookup for the bound segment this script will operate on. */
+    private SetOnce<LeafSearchLookup> leafLookup = new SetOnce<>();
+
+    /** A scorer that will return the score for the current document when the script is run. */
+    private SetOnce<Scorer> scorer = new SetOnce<>();
+
+    /**
+     * Construct an unbound search script.
+     * Calling any methods except {@link #forSegment(LeafReaderContext)} on the returned instance is undefined.
+     */
+    public SearchScript(Map<String, Object> params, SearchLookup lookup) {
+        this.params = params;
+        this.lookup = lookup;
+    }
+
+    /**
+     * Constructs a copy of this script that is bound to the given segment.
+     *
+     * @param leaf The reader context for a Lucene segment
+     * @return A new {@link SearchScript} instance
+     */
+    public SearchScript forSegment(LeafReaderContext leaf) {
```

Comment:
Script engines don't need to care about this. They have the option of overriding in order to optimize some things (as expressions do), but this should not be abstract, otherwise every script engine must specialize this, and understand SearchScript (or some other base class) specifically. The intent here is to eventually make script engines oblivious to the fact the SearchScript has this special behavior unlike other scripts.